[Proposal] Magic rework (Phase 1)

Xaver-DaRed opened the pull request:
<!-- remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm: -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] that I agree to LandSandBoat's [Limited Contributor License Agreement](https://github.com/LandSandBoat/server/blob/base/.github/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I have **read the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)**
- [eh...] that I've _**tested my code and things my code changed**_ since the last commit in the PR, and will test after any later commits

An atempt to make magic calculations sane, to put it simply.

- What this is:

Right now, spells are being handled in magic.lua, core and their own spell scripts.
This would be fine, but magic.lua has many problems.
The first being that it is very hard to follow, with the effect of making a correction to spells a lot harder than it should be.
The second is that it handles calculations chaotically and outright wrong, making magic not work at all as it should.

This attempts to make magic logic easier to follow by dividing the different magic processes into their own scripts, stripping logic from spell scripts in the process.

This also atempts to replicate retails damage calculations better than how we currently do it.

This is only the first phase, wich handles damage calculations. Second phase would handle enfeebling spells, and further phases other types of spells (healing, enhancing, etc...)

- What this isn't:

A replacement for magic.lua. Spells affected don't require it, but spells unaffected do.
The endgame would be not to need it,

A fix for every problem. Or for any problem. If it fixes something is by sheer coincidence/luck/mistake.

A finished product.

A fully tested product. I'm just one person. It's tested enough to know it compiles, it doesn't IMEDIATELY crash the server and spells do actual damage.

Xaver-DaRed committed:
Damage Calculation File

Just that file. It handles logic for:
- Most black mage damaging spells
- Ninja damaging Spells

Xaver-DaRed committed:
Black mage wind spells

New scripts for:
- Aera 1, 2 and 3
- Aeroga 4 and 5, with an anti-cheat check.
- Aero 6
Edits:
- Aero 1, 2, 3, 4 and 5
- Aeroga 1, 2 and 3
- Aeroja
- Tornado 1 and 2

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+    -- Get base damage (V in English wiki, D in JP wiki) depending on caster.
+    if caster:isPC() then
+        baseSpellDamage = damageSpellParams[spellId][6]
+    else
```
Just a suggestion, might want to define some constants to cover the magic numbers in most table references.  The arrays are pretty beefy, and that would go a long way for readability (this continues throughout the PR)

TeoTwawki replied:
like
```lua
local vPC = 4
--[[ stuff between]]
baseSpellDamage = damageSpellParams[spellId][vPC]
```
?

Xaver-DaRed replied:
yeah, thats exactly it.
Gonna do it RQ ans see what the new lua-checks return while im at it

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+            statDiffBonus = math.floor(50 * spellMultiplier0) + math.floor(50 * spellMultiplier50) + math.floor(100 * spellMultiplier100) + math.floor(200 * spellMultiplier200) + math.floor(300 * spellMultiplier300) + math.floor((statDiff - 400) * spellMultiplier400)
+        else -- It's over 500!
+            statDiffBonus = math.floor(50 * spellMultiplier0) + math.floor(50 * spellMultiplier50) + math.floor(100 * spellMultiplier100) + math.floor(200 * spellMultiplier200) + math.floor(300 * spellMultiplier300) + math.floor(400 * spellMultiplier400) +  math.floor((statDiff - 500) * spellMultiplier500)
+        end
```
Can we handle some of this through multiple lines?

Xaver-DaRed committed:
Black mage Ice spells

New scripts for:
- Blizzara 1, 2 and 3
- Blizzaga 4 and 5, with an anti-cheat check.
- Blizzard 6
Edits:
- Blizzard 1, 2, 3, 4 and 5
- Blizzaga 1, 2 and 3
- Blizzaja
- Freeze 1 and 2

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+        -- BLM Job Point: Manafont Elemental Magic Damage +3
+        if caster:hasStatusEffect(xi.effect.MANAFONT) then
+            baseSpellDamageBonus = baseSpellDamageBonus + (caster:getJobPointLevel(xi.jp.MANAFONT_EFFECT) * 3)
+        end
```
Should be fine with order of ops without the parens

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+end
+
+function calculateMTDR(caster, target, spell)
+    local MTDR    = 0 -- The variable we want to calculate.
```
Can there be a comment defining what MTDR is, or perhaps a change in naming convention?

Xaver-DaRed replied:
I followed BG wiki naming conventions, but you are right, I'll add a comment.
MTDR stands for "Multiple Target Damage Reduction" btw XD

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+
+    if eleStaffBonus > 0 then
+        eleStaffBonus = 1 + (eleStaffBonus * 0.05)
+    else
```
Unneeded parens

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+        -----------------------------------
+        -- WHM Job Points
+        if casterJob == xi.job.WHM then
+            magicAcc = magicAcc + caster:getJobPointLevel(xi.jp.WHM_MAGIC_ACC_BONUS)
```
This is more on me for the initial changes, but these can move to if/elseif, since mainjob won't be changing during cast

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+        -- magicAcc from Merits.
+        -----------------------------------
+        -- BLM Merits
+        if casterJob == xi.job.BLM and skillType == xi.skill.ELEMENTAL_MAGIC then
```
Same as above for job checks

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+                dayAndWeather = dayAndWeather + 0.10
+            end
+        end
+        if math.random() < 0.33 or caster:getMod(elementalObi[spellElement]) >= 1 or isHelixSpell == true then
```
Small styling preference on my end, if there's multiple conditionals, could you put a newline in between to easily differentiate between blocks?

TeoTwawki replied:
```lua
if
    math.random() < 0.33 or
    caster:getMod(elementalObi[spellElement]) >= 1 or
    isHelixSpell == true
then
```
is the format specified by the style guide when conditionals are to wide or confusing, iirc

ps we usual do not need `== true` for most things, unless they can be something that is not true/false (nil?)

Xaver-DaRed replied:
Ya, I didnt consider it too long, honestly. That + the repetition in the following lines made me think it was ok as it is...
If there were both "or" and "and", I would have done it already.
If you both agree to divide this (and the ones bellow by extension), I will do it, but i think it will only serve to bloat this particular chain.

TeoTwawki replied:
I think I misread what clay was asking for here before quoting styling info. I think he just wanted a blank line between the previous `end` and the next `if`.

Xaver-DaRed replied:
oh. I see... kk, easy enough

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,863 @@
+end
+
+function calculateTMDA(caster, target, spell, spellElement)
+    local TMDA = 1 -- The variable we want to calculate
```
Same with MTDR, TMDA could use a comment or naming that identifies what its purpose is

Xaver-DaRed committed:
Black mage Earth spells

New scripts for:
- Stonera 1, 2 and 3
- Stonega 5, with an anti-cheat check.
- Stone 6
Edits:
- Stone 1, 2, 3, 4 and 5
- Stonega 1, 2, 3 and 4
- Stoneja
- Quake 1 and 2

Xaver-DaRed committed:
Black mage Fire spells

New scripts for:
- Fira 1, 2 and 3
- Firaga 5, with an anti-cheat check.
- Fire 6
Edits:
- Fire 1, 2, 3, 4 and 5
- Firaga 1, 2, 3 and 4
- Firaja
- Flare 1 and 2

Xaver-DaRed committed:
Black mage Thunder spells

New scripts for:
- Thundara 1, 2 and 3
- Thundaga 4 and 5, with an anti-cheat check.
- Thunder 6
Edits:
- Thunder 1, 2, 3, 4 and 5
- Thundaga 1, 2 and 3
- Thundaja
- Burst 1 and 2

Xaver-DaRed committed:
Black mage Water spells

New scripts for:
- Watera 1, 2 and 3
- Waterga 5, with an anti-cheat check.
- Water 6
Edits:
- Water 1, 2, 3, 4 and 5
- Waterga 1, 2, 3 and 4
- Waterja
- Flood 1 and 2

Xaver-DaRed committed:
Corrections to main file.

- Add Ninja spells merits, wich I forgot.
- Adress most coments.

Xaver-DaRed committed:
Add ninja spells and comet

claywar commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,883 @@
+            magicAcc = magicAcc + caster:getMerit(xi.merit.MAGICAL_ACCURACY)
+        end
+        -- OTHER merits. ???? which are...? Modifiers need a rework/reorganization
+        magicAcc = magicAcc + caster:getMerit(xi.merit.MAGIC_ACCURACY)
```
Where is this generic `xi.merit.MAGIC_ACCURACY` coming from?

Xaver-DaRed replied:
As I pointed out in the comment, I have no clue XD
It was there in the original magic.lua file and decided to keep it in case someone had more info on it.
It is on my "to destroy if no info is provided" private list

TeoTwawki replied:
its a modifier set by reading the [merit table](https://github.com/LandSandBoat/server/blob/base/sql/merits.sql#L226). Whether or not you can [actually merit it ](https://github.com/LandSandBoat/server/blob/base/sql/char_merit.sql)is another story.

The code fr applying merits is rather generic so you won't find core code doing application of the specific merit.

Xaver-DaRed replied:
oh, I see. It's RDM category 2 merit

TeoTwawki commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,883 @@
+
+-- Multiple target spells:
+-- Structure:           [spellId] = {  Stat used,                    Skill,                    Element, vNPC, mNPC,  vPC,   M0,  M50,  M100, M200, M300, M400, M500 },
+    [xi.magic.spell.AEROGA      ] = { xi.mod.INT, xi.skill.ELEMENTAL_MAGIC,      xi.magic.element.WIND,   93,    1,  100,  2.6,  1.8,     1,    0,    0,    0,    0 },
```
Maybe we could make these liens less wide by casting some things to variables or by getting some things from other locations rather than having to repeat them into the table/array

like
```lua
local INT = xi.mod.INT
local elemSkill = xi.skill.ELEMENTAL_MAGIC
```
```lua
[xi.magic.spell.AEROGA      ] = { INT, elemSklill,      xi.magic.element.WIND,   93,    1,  100,  2.6,  1.8,     1,    0,    0,    0,    0 },
```
or we use `spell:getSkillType()` to not need a table entry at all

TeoTwawki replied:
we can "get" the element as well.

Xaver-DaRed replied:
The reason I defined elements (and skills) into the table was to make less calls to external functions. I'm guessing it is making a call to the database.
I was hoping that someone (not me, someone that actually knows how to do it) would compare both methods and check wich one was more eficient resource wise.
It also helped me when filling the tables. A visual help, so to speak.

I'm not oposed to creating local vars to shorten the length though.

zach2good replied:
I see Teo has already had the same idea as me...

zach2good commented:
I don't have time to look at this for a few days, but a 90-second high level skim leaves me very impressed. I thought I would hate the giant table of spell parameters but it looks pretty good!

I'll try to get a review in at the weekend 👍 

I don't know if this would make the information too dense, but you can reduce how wide your tables are like this:
```lua
-- Assign a shorter name
local ma = xi.magic
local we = xi.weather
local ef = xi.effect
local mod = xi.mod

-- Big structure table
```

This will work as long as the values aren't being written to, I'm not sure if this operates by value or reference 🤷

zach2good commented:
One of the strengths of this design is how easy it is for people who want to use era-style values/calculations.
If the big tables of params were changes to be attached to the `xi` global:

```
local damageSpellParams = ...
```
->
```
xi.magic_utils.damageSpellParams = ...
``

`modules/era_spell_values.lua` ->
```lua
local module = require("scripts/modules/module")
module.override(xi.magic_utils, "damageSpellParams", {
    -- Table of era-style values
}
)
```

... or `spell_damage` can be overridden,

TeoTwawki commented:
> 
> 
> One of the strengths of this design is how easy it is for people who want to use era-style values/calculations.
> If the big tables of params were changes to be attached to the `xi` global:
> 
> ```lua
> local damageSpellParams = ...
> ```
> 
> ->
> 
> ```lua
> xi.magic_utils.damageSpellParams = ...
> ```
> 
> `modules/era_spell_values.lua` ->
> 
> ```lua
> local module = require("scripts/modules/module")
> module.override(xi.magic_utils, "damageSpellParams", {
>     -- Table of era-style values
> }
> )
> ```
> 
> ... or `spell_damage` can be overridden,

its also turns out that monsters are still using those era calculations too so now we have a retail accuracy reason to continue supporting them

Xaver-DaRed committed:
Changes to main file.

- Add support for old school calculations.
- Add support for divine magic.
- Address several comments.

Xaver-DaRed committed:
Fix Derp.

TeoTwawki commented on scripts/globals/spells/waterga_v.lua:
```diff
@@ -0,0 +1,21 @@
+local spell_object = {}
+
+spell_object.onMagicCastingCheck = function(caster, target, spell)
+    if caster:isPC() then -- Mob Only Spell.
```
this isn't necessary. a player normally can't obtain it to cast it, we do not need to block it. If you choose to GM yourself the spell to get it, its not a bug its just not the projects problem either.

zach2good commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,908 @@
+        -- Ninja Category 2 merits.
+        mab = mab + caster:getMerit(xi.merit.NIN_MAGIC_BONUS)
+        -- Ninja Category 1 merits
+        if spellId >= 320 and spellId <= 322 then     -- Katon series.
```
You should be able to use `xi.magic.spellFamily` for these checks:
https://github.com/LandSandBoat/server/blob/base/scripts/globals/spell_data.lua#L113

For checks (like ancient magic) I wouldn't be against helpers living somewhere so you can just call `xi.spell.isAncientMagic(spellId)`. Even if you've commented it well

ShiyoKozuki replied:
Few things:
**Elemental Seal:**
From personal testing(No proof, sorry), elemental seal seems to make your resist go up a tier. So if you have 50% SDT, it goes to from 1/2 to 1. If you have sub 50% SDT, it goes from 1/4 to 1/2. 

**Resist stats for players only**
I have testing for this and proof!
https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/107/#3570888

https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/109/#3571304

```
Enfeebling test on a 2 resist state enfeeble, via having level ~70 Colibris reflect Dispel. I had Saboteur Addle II on the mob and 4/5 Malignance, so MEVA should be capped.

Neutral dark resistance: 1 landed, 9 resisted
-3 dark resistance: 10 landed
```

So, basically..

If you have any negative resists to the element hitting you, it will always land for 1/2 resist minimum(including enfeebles)
If you do not have any positive resist to the element hitting you, you cannot resist to 1/8.

Xaver-DaRed committed:
Main file Adjustments

- Fix undead bonus function
- Fix table references
- Fix math, becouse math is hard

Xaver-DaRed committed:
Resolve conversations

Xaver-DaRed committed:
Main file edits and trigger new CI

Bring it

Xaver-DaRed committed:
Test CI changes

Xaver-DaRed committed:
Correct most (...) CI fails

Xaver-DaRed committed:
Fix variable

Xaver-DaRed commented:
Ok, thats gonna need to be squashed...

Finished fixing most warnings given by CI.
Warnings in spell scripts are out of the scope of this PR. Mainly becouse they involve phase 2 anyway (rewrite debuff logic)
Warnings in main file... well, maybe i could do something about those, but honestly, they involve a different file altoguether and If I dare to open it, I would have to rewrite it too.
I refuse to, at this time XD

Xaver-DaRed committed:
Simplify MTDR

Xaver-DaRed committed:
Make separate file for spell stat tables. Add Divine nukes

kaincenteno commented:
Love the param table, so easy to read than going through each spell. Def an improvement :smile:

ShiyoKozuki commented:
Few things:
Elemental Seal:
From personal testing(No proof, sorry), elemental seal seems to make your resist go up a tier. So if you have 50% SDT, it goes to from 1/2 to 1. If you have sub 50% SDT, it goes from 1/4 to 1/2.

Resist stats for players only
I have testing for this and proof!
https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/107/#3570888

https://www.ffxiah.com/forum/topic/46016/first-and-final-line-of-defense-v20/109/#3571304

```
Enfeebling test on a 2 resist state enfeeble, via having level ~70 Colibris reflect Dispel. I had Saboteur Addle II on the mob and 4/5 Malignance, so MEVA should be capped.

Neutral dark resistance: 1 landed, 9 resisted
-3 dark resistance: 10 landed
So, basically..
```

If you have any negative resists to the element hitting you, it will always land for 1/2 resist minimum(including enfeebles)
If you do not have any positive resist to the element hitting you, you cannot resist to 1/8.

TeoTwawki commented:
> When I played SCH at 99 cap(not ilvl), I needed obscene macc to land stun reliably on Legion enemies with high SDT to thunder and high mevasion.

you should know that Legion had modified stun resistances all over creation because some of the admins didn't want it being too reliable to "just throw more blm or blue at". drk was collateral damage. A lot of (but not all of) that nonsense was stripped out (by me) prior to their public code dump when they shut down.

Xaver-DaRed committed:
Merge branch 'base' into proposal/magic-rework

Xaver-DaRed committed:
Update for settings.lua file changes

Xaver-DaRed committed:
Merge branch 'proposal/magic-rework' of https://github.com/Xaver-DaRed/server into proposal/magic-rework

TeoTwawki commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,790 @@
+            dayAndWeather = dayAndWeather + 0.10
+        end
+
+    elseif weather == xi.magic.singleWeatherWeak[spellElement] then
```
possibly to table/loop much if this if/else tower? it looks like a lot of repeating code goes on here.

TeoTwawki commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -0,0 +1,790 @@
+    local ancientMagicBurstBonus          = 0
+    local skillchainTier, skillchainCount = FormMagicBurst(spellElement, target) -- External function. Not present in magic.lua.
+
+    if spellId >= 204 and spellId <= 215 then -- If spell is Ancient Magic
```
I think we have spell names enum'd somewhere now so that we don't have to have the ID use

Xaver-DaRed committed:
Update to TMDA

Xaver-DaRed committed:
Fix derp

Xaver-DaRed committed:
Cleanup TMDA

Xaver-DaRed committed:
Reorganize weather a little

TeoTwawki committed:
convert magic numbers to human readable

add TODO regarding spell ID and spell family ID tabling

spellcheck and other minor comment adjustments

add small reminder for future sdt and resist related things (always be clear which one we are working with)

TeoTwawki committed:
comment tweakage

TeoTwawki committed:
Merge branch 'base' of https://github.com/LandSandBoat/server into proposal/magic-rework

Xaver-DaRed committed:
Correct Settings

ShiyoKozuki commented:
> > When I played SCH at 99 cap(not ilvl), I needed obscene macc to land stun reliably on Legion enemies with high SDT to thunder and high mevasion.
> 
> you should know that Legion had modified stun resistances all over creation because some of the admins didn't want it being too reliable to "just throw more blm or blue at". drk was collateral damage. A lot of (but not all of) that nonsense was stripped out (by me) prior to their public code dump when they shut down.
> 
> not saying there isn't anything else, but thought was worth the mention/grain of salt on using legion as the example.

Legion meaning the end game event, not the server.

Xaver-DaRed committed:
Merge branch 'LandSandBoat:base' into proposal/magic-rework

Xaver-DaRed committed:
Apply SDT, elemental resistances and corrected mods and unused variables

Xaver-DaRed committed:
Lower complexity in resistance function. Fix table leftovers.

Xaver-DaRed committed:
Fix trailing whitespace

TeoTwawki commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -43,8 +43,7 @@ local blmMerit               = { xi.merit.FIRE_MAGIC_POTENCY,  xi.merit.ICE_MAGI
 -- Table variables.
-local spellTable   = xi.magic_utils.parameters.damageParams
-local elementTable = xi.magic_utils.parameters.elementParams
+local table = xi.magic_utils.parameters.damageParams
```
bad variable name, avoid using keywords as names

Xaver-DaRed committed:
Tested and fixes related to TMDA and bad Mod conversion on my end.

Xaver-DaRed commented:
Went through and tested this.
- Nothing crashes or errors IMMEDIATELY.
- Magic Bursts work and modifiers are applied.
- Resists happen.
- SDT seems to happen aswell.

TeoTwawki commented on scripts/globals/magic_utils/spell_damage.lua:
```diff
@@ -206,7 +206,10 @@ end
+    local SDT    = 1 -- The variable we want to calculate
+    local SDTMod = target:getMod(xi.magic.specificDmgTakenMod[spellElement]) -- Elemental SDT mods are Base 100.
+
+    SDT = (SDTMod / -100) + 1 -- Convert Modifier value.
```
~~like I was saying on discord, the sql has to do that to create modifiers. you do not want to repeat/reverse it and they are -not- base 100 they are base 1000 to support half percent~~

From modifier.h and status.lua:
```
Example: 1000 = 100%, 875= 87.5%
```
~~this confusion is precisely why I left that comment there~~

Just realized everything wasn't fully converted script side. will take care of this when it is

Xaver-DaRed committed:
Add comment/reminder concerning SDT modifiers conversion.

TeoTwawki reviewed (approved):
Outside of an expected SDT [issue](https://github.com/LandSandBoat/server/issues/1046) we can handle after, this looks good to me.

Era-Lusiphur reviewed (approved):
Did a good bit of before/after testing, looked at results with @TeoTwawki. Everything appears to have results with an expected SDT skew Teo was anticipating. 

Tested blue magic, divine, elemental, enfeebling, and songs.
